File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/grammar-graph/get-input-as-code-points.ts
import { ValueError } from "../utils/errors/python-errors.js";
import type { ValidInput } from "./grammar-graph-types.js";

export const getCodePoint = (char: string): number => {
  const codePoint = char.codePointAt(0);
  if (codePoint === undefined) {
    throw new ValueError(`Could not get code point for character: ${char}`);
  }
  if (!Number.isInteger(codePoint)) {
    throw new ValueError("code_point must be an integer!");
  }
  return codePoint;
};

export const getInputAsCodePoints = (src: ValidInput): number[] => {
  if (typeof src === "number") {
    return [src];
  }

  if (Array.isArray(src)) {
    for (const c of src) {
      if (!Number.isInteger(c)) {
        throw new ValueError(
          `code_point must be an integer for ${c} if src is a list`,
        );
      }
    }
    return src;
  }

  if (typeof src === "string") {
    // Python iterates strings by code point; `Array.from` does the same.
    return Array.from(src).map(getCodePoint);
  }

  throw new ValueError(`Invalid input type: ${typeof src}`);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T150124Z_a7cd2ee9/20260910T084451Z_b3565ede/ported_implementation/gbnf/grammar_graph/get_input_as_code_points.py
from .grammar_graph_types import ValidInput


def get_code_point(char: str) -> int:
    if not char:
        raise ValueError(f"Could not get code point for character: {char}")
    return ord(char[0])


def get_input_as_code_points(src: ValidInput) -> list[int]:
    if isinstance(src, bool):
        raise TypeError(f"Invalid input type: {type(src)}")

    if isinstance(src, int):
        return [src]

    if isinstance(src, list):
        for c in src:
            if not isinstance(c, int) or isinstance(c, bool):
                raise TypeError(
                    f"code_point must be an integer for {c} if src is a list"
                )
        return src

    if isinstance(src, str):
        return [get_code_point(char) for char in src]

    raise TypeError(f"Invalid input type: {type(src)}")
